FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpDecrypt.kt

```diff
@@ -0,0 +1,100 @@
+ * Contributors: DenBond7
+ */
+
+package com.flowcrypt.email.security.pgp
+
+import com.flowcrypt.email.util.exception.DecryptionException
+import org.bouncycastle.openpgp.PGPDataValidationException
+import org.bouncycastle.openpgp.PGPException
+import org.bouncycastle.openpgp.PGPSecretKeyRingCollection
+import org.pgpainless.PGPainless
+import org.pgpainless.decryption_verification.OpenPgpMetadata
+import org.pgpainless.exception.MessageNotIntegrityProtectedException
+import org.pgpainless.exception.ModificationDetectionException
+import org.pgpainless.key.protection.SecretKeyRingProtector
+import java.io.InputStream
+import java.io.OutputStream
+
+/**
+ * @author Denis Bondarenko
+ *         Date: 5/11/21
+ *         Time: 2:10 PM
+ *         E-mail: DenBond7@gmail.com
+ */
+object PgpDecrypt {
+  val DETECT_SEPARATE_ENCRYPTED_ATTACHMENTS_PATTERN =
+    "(?i)(\\.pgp$)|(\\.gpg$)|(\\.[a-zA-Z0-9]{3,4}\\.asc$)".toRegex()
+
+  fun decrypt(
+    srcInputStream: InputStream,
+    destOutputStream: OutputStream,
+    pgpSecretKeyRingCollection: PGPSecretKeyRingCollection,
+    protector: SecretKeyRingProtector
+  ): OpenPgpMetadata {
+    srcInputStream.use { srcStream ->
+      destOutputStream.use { outStream ->
+        try {
+          val decryptionStream = PGPainless.decryptAndOrVerify()
+            .onInputStream(srcStream)
+            .decryptWith(protector, pgpSecretKeyRingCollection)
+            .doNotVerify()
+            .build()
+
+          decryptionStream.use { it.copyTo(outStream) }
+          return decryptionStream.result
+        } catch (e: Exception) {
+          throw processDecryptionException(e)
+        }
+      }
+    }
+  }
+
+  private fun processDecryptionException(e: Exception): Exception {
+    return when (e) {
+      is PGPException -> {
+        when {
+          e.message == "checksum mismatch at 0 of 20" -> {
+            DecryptionException(DecryptionErrorType.WRONG_PASSPHRASE, e)
+          }
+
+          e.message?.contains("exception decrypting session info") == true
+              || e.message?.contains("encoded length out of range") == true
+              || e.message?.contains("Exception recovering session info") == true
+              || e.message?.contains("No suitable decryption key") == true -> {
+            DecryptionException(DecryptionErrorType.KEY_MISMATCH, e)
+          }
+
+          else -> DecryptionException(DecryptionErrorType.OTHER, e)
+        }
+      }
+
+      is MessageNotIntegrityProtectedException -> {
+        DecryptionException(DecryptionErrorType.NO_MDC, e)
+      }
+
+      is ModificationDetectionException -> {
+        DecryptionException(DecryptionErrorType.BAD_MDC, e)
+      }
+
+      is PGPDataValidationException -> {
+        DecryptionException(DecryptionErrorType.KEY_MISMATCH, e)
+      }
+
+      is DecryptionException -> e
+
+      else -> DecryptionException(DecryptionErrorType.OTHER, e)
+    }
+  }
+
+  enum class DecryptionErrorType {
+    KEY_MISMATCH,
```

Comment:
I think we must already have such enum in the PgpMsg. Can it be used here?

Reply:
Currently, we have the following:

```
PgpArmor - is responsible for armor source
PgpDecrypt - is responsible for decrypting things
PgpEncrypt - is responsible for encrypting things
PgpKey - is responsible for handling keys
PgpMsg - is responsible for processing PGP messages
PgpPwd - is responsible for passphrases
```

In my opinion, it's more transparent to store the decryption error types in `PgpDecrypt`. Because you work on #1057, #1223 I've decided don't remove that code to prevent conflicts. We can clean code in a separate issue.

Reply:
It's not obvious decision, I suggest to move it now. I am ok with the fact it may require then to resolve some conflicts.